misoc/targets/kasli.py

```diff
@@ -34,25 +36,61 @@ def __init__(self, cd, async_reset):
         ]
 
 
-class _CRG(Module):
-    def __init__(self, platform):
+class _CRG(Module, AutoCSR):
+    def __init__(self, platform, freq=125.0e6):
         self.clock_domains.cd_sys = ClockDomain()
         self.clock_domains.cd_sys4x = ClockDomain(reset_less=True)
         self.clock_domains.cd_sys4x_dqs = ClockDomain(reset_less=True)
         self.clock_domains.cd_clk200 = ClockDomain()
 
-        clk125 = platform.request("clk125_gtp")
-        platform.add_period_constraint(clk125, 8.)
-        self.clk125_buf = Signal()
-        self.clk125_div2 = Signal()
+        self.clock_sel = CSRStorage()
+        self.pll_locked = CSRStatus()
+        self.mmcm_locked = CSRStatus()
+
+        # old "sys" clock, now bootstrap
+        bootstrap = platform.request("clk125_gtp")
+        platform.add_period_constraint(bootstrap, 8.)
+        bootstrap_buf = Signal()
+        bootstrap_div2 = Signal()
         self.specials += Instance("IBUFDS_GTE2",
             i_CEB=0,
-            i_I=clk125.p, i_IB=clk125.n,
-            o_O=self.clk125_buf,
-            o_ODIV2=self.clk125_div2)
+            i_I=bootstrap.p, i_IB=bootstrap.n,
+            o_O=bootstrap_buf,
+            o_ODIV2=bootstrap_div2)
+
+        # "rtio" clock (mgt flavor)
+        if platform.hw_rev == "v2.0":
+            si5324_out = platform.request("cdr_clk_clean")
+            
+        else:
+            si5324_out = platform.request("si5324_clkout")
+
+        if freq == 125.0e6:
+            si5324_period = 8.
+            pll_mult = 16
+        elif freq == 100.0e6:
+            si5324_period = 10.
+            pll_mult = 20
+        else:
+            raise NotImplementedError
+         
+        self.si5324_buf = Signal()
+        si5324_div2 = Signal()
+
+        platform.add_period_constraint(si5324_out, si5324_period)
+
+        self.specials += Instance("IBUFDS_GTE2",
+            i_CEB=0,
+            i_I=si5324_out.p, i_IB=si5324_out.n,
```

Comment:
Should be called ``cdr_clk_...`` since it may not come from the Si5324 after WRPLL.